app/javascript/components/widget-controller.jsx

```diff
@@ -33,234 +30,59 @@ const getWidgets = gql`
-    }
-  };
-
-  render() {
-    const { currentWidgetId, pausedTime } = this.state;
-
-    const fetchPolicy = currentWidgetId === 0 ? 'network-only' : 'cache-first';
-
-    const { cityName } = this.props;
-
-    return (
-      <Query
-        query={getWidgets}
-        variables={{ id: currentWidgetId, cityName }}
-        fetchPolicy={fetchPolicy}
-        onCompleted={response => {
-          const { location } = response;
-          const { widgets } = location;
-          const { enabled: enabledWidgets, byIdOrFirst: current } = widgets;
-
-          const index = enabledWidgets.findIndex(w => w.id === current.id);
-
-          const prevWidgetId =
-            enabledWidgets[mathMod(index - 1, enabledWidgets.length)].id;
-          const nextWidgetId =
-            enabledWidgets[mathMod(index + 1, enabledWidgets.length)].id;
-
-          this.setState({
-            currentWidgetId: current.id,
-            nextWidgetId,
-            prevWidgetId,
-            transitionTime: Date.now() + current.durationSeconds * 1000,
-          });
-        }}
-      >
-        {({ loading, error, data }) => {
-          if (loading) {
-            return (
-              <Wrapper>
-                <FullPanel>
-                  <LoadingMessage />
-                </FullPanel>
-                <SidePanel
-                  widgets={[]}
-                  selectedWidgetId={0}
-                  showWeather={false}
-                  totalTime={0}
-                  isPaused={false}
-                  cityName={cityName}
-                />
-              </Wrapper>
-            );
-          }
-
-          if (error) {
-            // eslint-disable-next-line
-            console.error(error);
-            return (
-              <Wrapper>
-                <FullPanel>
-                  <DisconnectedMessage />
-                </FullPanel>
-                <SidePanel
-                  widgets={[]}
-                  selectedWidgetId={0}
-                  showWeather={false}
-                  totalTime={0}
-                  isPaused={false}
-                  cityName={cityName}
-                />
-              </Wrapper>
-            );
-          }
-
-          const { location } = data;
-          const { widgets } = location;
-          const { enabled: enabledWidgets, byIdOrFirst: current } = widgets;
+  return (
+    <Query
+      query={getWidgets}
+      variables={{ id: newWidgetId, cityName }}
+      fetchPolicy={fetchPolicy}
+      onCompleted={response => {
+        const {
+          location: {
+            widgets: { byIdOrFirst: current },
+          },
+        } = response;
+        setNewWidgetId(current.id);
+      }}
+    >
+      {({ loading, error, data }) => {
+        if (loading) {
+          return <LoadingDisplay cityName={cityName} />;
+        }
 
-          const WidgetElement = widgetElements[current.name].Panel;
+        if (error) {
+          // eslint-disable-next-line
+          console.error(error);
```

Comment:
Last week we discussed that your refactoring may not cause much of an issue, but I think this is something you and I should go over again after you merge since side-panel is no longer imported, causing a difference in the functionality of error and loading.

Reply:
Yeah the SidePanel was moved to the WidgetDisplay - we should definitely coordinate resolving any merge conflicts at some point.